File: 0200-number-of-islands/0200-number-of-islands.java
class Solution {
    int[][] heads = {{0, 1}, {0, -1}, {-1, 0}, {1, 0}};
    
    public int numIslands(char[][] grid) {
        int m = grid.length;
        int n = grid[0].length;
        
        int count = 0;
        for(int i = 0; i < m; i++) {
            for(int j = 0; j < n; j++) {
                // 섬의 시작 지점인 경우 
                if(grid[i][j] == '1') {
                    count++;
                    
                    // bfs 탐색을 위함
                    Queue<int[]> queue = new LinkedList<>();
                    queue.add(new int[]{i, j});
                    
                    // 방문 처리
                    grid[i][j] = '0';
                    
                    // BFS 탐색
                    while(!queue.isEmpty()) {
                        int[] pos = queue.poll();
                        int x = pos[1];
                        int y = pos[0];
                        
                        for(int[] head : heads) {
                            int newx = x + head[1];
                            int newy = y + head[0];
                            
                            // 인접한 땅을 큐에 추가하고 방문 처리
                            if(newy >= 0 && newx >= 0 && newy < m && newx < n && grid[newy][newx] == '1') {
                                queue.add(new int[]{newy, newx});
                                grid[newy][newx] = '0';
                            }
                        }
                    }
                }
            }
        }
        
        return count;
    }
}

File: 0200-number-of-islands/0200-number-of-islands.kt
class Solution {
    
    fun numIslands(grid: Array<CharArray>): Int {
        fun dfs(i: Int, j: Int) {
            // 현재 위치가 그리드 밖이거나 물(0)인 경우 탐색 종료
            when {
                (i < 0) ||
                (i >= grid.size) ||
                (j < 0) ||
                (j >= grid[0].size) ||
                (grid[i][j] == '0') -> return
            }
            
            // 방문 지점 마킹
            grid[i][j] = '0'
            
            // dfs 탐색
            dfs(i, j+1)
            dfs(i, j-1)
            dfs(i+1, j)
            dfs(i-1, j)
        }
        
        var count = 0
        for(i in grid.indices) {
            for(j in grid[i].indices) {
                // 육지(1)인 경우 dfs 탐색 시작
                if(grid[i][j] == '1') {
                    dfs(i, j)
                    
                    // dfs 탐색이 끝나면 섬 하나를 찾았으므로 카운트를 1만큼 증가
                    count++
                }
            }
        }
        
        // 섬의 개수 반환
        return count
        
    }
}